### PR TITLE
Refactor search_and_install_packs to use networkX for dependencies

### PR DESCRIPTION
**Pull request by mmhw**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/CIAC-9115)

## Description
- Refactored search_and_install_packs to use networkX directed graph to model dependencies between packs
- Handle cycles in dependency graph by merging cyclic nodes
- Topologically sort packs to install respecting dependencies

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Tests/Marketplace/search_and_install_packs.py [hunk 1]
```diff
@@ -742,82 +561,297 @@ def search_and_install_packs_and_their_dependencies_private(test_pack_path: str,
...
+    Returns:
+        dict[str, str]: A dictionary mapping the original node names that were part of 
+        cycles to the merged node name for that cycle.
+    """
+    logging.debug(
+        f"Found the following cycles in the graph: {list(nx.simple_cycles(graph))}"
+    )
+    map_cycles = {}
+    while list(nx.simple_cycles(graph)):
+        cycle = list(nx.simple_cycles(graph))[0]
```
Comment (DeanArbel) [inline Tests/Marketplace/search_and_install_packs.py:605]: this can be simplified 
```suggestion
    while cycles := list(nx.simple_cycles(graph)):
        cycle = cycles[0]
```

**Comment by mmhw** [comment]
[Content Nightly](https://gitlab.xdr.pan.local/xdr/cortex-content/content/-/pipelines/775143)

File: pyproject.toml [hunk 2]
```diff
@@ -156,6 +156,7 @@ tabulate = ">=0.8.9,<0.10.0"
 jira = "^3.5.2"
 junitparser = "^3.1.0"
 tenacity = "^8.2.3"
+pygraphviz = "*"
```
Comment (dorschw) [inline pyproject.toml:159]:
```suggestion
pygraphviz = "^1.12"
```

Comment (dorschw) [inline pyproject.toml:159]: is this dependency required for CI to run, or just locally?

File: Tests/Marketplace/search_and_install_packs.py [hunk 3]
```diff
@@ -742,82 +564,298 @@ def search_and_install_packs_and_their_dependencies_private(test_pack_path: str,
...
+    Returns:
+        The JSON response containing the packs dependencies for the given page
+    """
+    api_endpoint = "/contentpacks/marketplace/search"
+    body = {
+        "page": page,
+        "size": PAGE_SIZE_DEFAULT,
+        "sort": [
+            {"field": "searchRank", "asc": False},
+            {"field": "updated", "acs": False},
```
Comment (yaakovpraisler) [inline Tests/Marketplace/search_and_install_packs.py:691]:
```suggestion
            {"field": "updated", "asc": False},
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 4]
```diff
@@ -842,3 +879,122 @@ def create_batches(list_of_packs_and_its_dependency: list):
...
+    Searches for the packs from the specified list, searches their dependencies, and then
+    installs them.
+
+    Args:
+        pack_ids (list): A list of the pack ids to search and install.
+        client (demisto_client): The client to connect to.
+        hostname (str): Hostname of instance. Using for logs.
+        multithreading (bool): Whether to use multithreading to install packs in parallel.
+            If multithreading is used, installation requests will be sent in batches of each pack and its dependencies.
+        production_bucket (bool): Whether the installation is in post update mode. Defaults to False.
```
Comment (yaakovpraisler) [inline Tests/Marketplace/search_and_install_packs.py:923]:
```suggestion
        production_bucket (bool): Whether the installation is in post update mode. Default is True.
```

File: pyproject.toml (hunk 2 above)
Comment (mmhw) [reply pyproject.toml]: It required for CI

File: Tests/Marketplace/search_and_install_packs.py (hunk 3 above)
Comment (mmhw) [reply Tests/Marketplace/search_and_install_packs.py]: I removed it completely

File: Tests/configure_and_test_integration_instances.py [hunk 5]
```diff
@@ -345,7 +345,7 @@ def get_changed_integrations(self, packs_not_to_install: set[str] | None = None)
     def concurrently_run_function_on_servers(self, function=None, pack_path=None, service_account=None) -> tuple[bool, list[Any]]:
         raise NotImplementedError
 
-    def install_packs(self, pack_ids: list | None = None, multithreading=True, production_bucket: bool = True) -> bool:
+    def install_packs(self, pack_ids: list | None = None, install_packs_in_batches=False, production_bucket: bool = True) -> bool:
```
Comment (daryakoval) [inline Tests/configure_and_test_integration_instances.py:348]: Hi,
Why this change is necessary?

Comment (mmhw) [reply Tests/configure_and_test_integration_instances.py:348]: Because following my changes multithreading is not used, and the only difference is whether to install in batches or not.

**Review by DeanArbel — APPROVED** [review]
Very well done
